src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -95,56 +95,77 @@ EnsureReferenceTablesExistOnAllNodes(void)
 void
 EnsureReferenceTablesExistOnAllNodesExtended(char transferMode)
 {
+	List *referenceTableIdList = NIL;
+	uint64 shardId = INVALID_SHARD_ID;
+	List *newWorkersList = NIL;
+	const char *referenceTableName = NULL;
+	int colocationId = CreateReferenceTableColocationId();
+
 	/*
-	 * Prevent this function from running concurrently with itself.
-	 *
-	 * It also prevents concurrent DROP TABLE or DROP SCHEMA. We need this
-	 * because through-out this function we assume values in referenceTableIdList
-	 * are still valid.
+	 * Most of the time this function should result in a conclusion where we do not need
+	 * to copy any reference tables. To prevent excessive locking the majority of the time
+	 * we run our precondition checks first with a lower lock. If, after checking with the
+	 * lower lock, that we might need to copy reference tables we check with a more
+	 * aggressive and self conflicting lock. It is important to be self conflicting in the
+	 * second run to make sure that two concurrent calls to this routine will actually not
+	 * run concurrently after the initial check. That is also the reason why we release
+	 * the lock between the two iterations of precondition checks.
 	 *
-	 * We don't need to handle other kinds of reference table DML/DDL here, since
-	 * master_copy_shard_placement gets enough locks for that.
+	 * If after two iterations of precondition checks we still find the need for copying
+	 * reference tables we exit the forloop with the last lock held. This will prevent
+	 * concurrent DROP TABLE and create_reference_table calls so that the list of
+	 * reference tables we operate on are stable.
 	 *
-	 * We also don't need special handling for concurrent create_refernece_table.
-	 * Since that will trigger a call to this function from another backend,
-	 * which will block until our call is finished.
+	 * Since the changes to the reference table placements are made via loopback
+	 * connections we release the final lock held at the end of this function.
 	 */
-	int colocationId = CreateReferenceTableColocationId();
-	LockColocationId(colocationId, ExclusiveLock);
-
-	List *referenceTableIdList = CitusTableTypeIdList(REFERENCE_TABLE);
-	if (referenceTableIdList == NIL)
+	LOCKMODE lockmodes[] = { AccessShareLock, ExclusiveLock };
+	for (int l = 0; l < lengthof(lockmodes); l++)
```

Comment:
This for loop seems hard to digest. Do we really need that? Can't we do something like here: https://github.com/citusdata/citus/compare/fix/reference-lock-contention...onder/fix/reference-lock-contention?expand=1

Where, we gradually increase the lock levels? I see that one isolation test fails on that branch, need to look into that -- could be something minor or I miss something that this loop does

Reply:
Problem there is that before acquiring the ExlusiveLock we run into a situation where the table containing `shardId` could have been dropped due to `ShareLock` not protecting on membership changes on the colocation group. Since that `shardId` further down is used to copy the reference tables it would cause very hard to understand error messages to the user.

I have tried many different approaches, including the extraction of the check into a function we could call twice. However, due to the checks storing some information that will be used throughout the rest it becomes very convoluted to get these variables populated.

Also, you never release the sharelock on the colocation id, which leaves room for some deadlock potential I believe of concurrently running `EnsureReferenceTablesExistOnAllNodes`.